applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp

```diff
@@ -1517,45 +1508,31 @@ void SmallStrainUPwDiffOrderElement::InitializeNodalVariables(ElementVariables&
     KRATOS_CATCH("")
 }
 
-double SmallStrainUPwDiffOrderElement::CalculateBiotCoefficient(const ElementVariables& rVariables,
-                                                                const bool& hasBiotCoefficient) const
+std::vector<double> SmallStrainUPwDiffOrderElement::CalculateBiotCoefficients(const std::vector<Matrix>& rConstitutiveMatrices) const
 {
-    KRATOS_TRY
-
-    const PropertiesType& rProp = this->GetProperties();
-
-    // Properties variables
-    if (hasBiotCoefficient) {
-        return rProp[BIOT_COEFFICIENT];
-    } else {
-        // calculate Bulk modulus from stiffness matrix
-        return 1.0 - CalculateBulkModulus(rVariables.ConstitutiveMatrix) / rProp[BULK_MODULUS_SOLID];
-    }
+    std::vector<double> result;
+    std::transform(rConstitutiveMatrices.begin(), rConstitutiveMatrices.end(),
+                   std::back_inserter(result), [this](const Matrix& rConstitutiveMatrix) {
+        return CalculateBiotCoefficient(rConstitutiveMatrix);
+    });
 
-    KRATOS_CATCH("")
+    return result;
 }
 
-void SmallStrainUPwDiffOrderElement::InitializeBiotCoefficients(ElementVariables& rVariables,
-                                                                const bool& hasBiotCoefficient)
+double SmallStrainUPwDiffOrderElement::CalculateBiotCoefficient(const Matrix& rConstitutiveMatrix) const
 {
     KRATOS_TRY
 
     const PropertiesType& rProp = this->GetProperties();
 
-    rVariables.BiotCoefficient = CalculateBiotCoefficient(rVariables, hasBiotCoefficient);
-
-    if (rProp[IGNORE_UNDRAINED]) {
-        rVariables.BiotModulusInverse =
-            (rVariables.BiotCoefficient - rProp[POROSITY]) / rProp[BULK_MODULUS_SOLID] + rProp[POROSITY] / TINY;
+    // Properties variables
+    if (rProp.Has(BIOT_COEFFICIENT)) {
+        return rProp[BIOT_COEFFICIENT];
     } else {
-        rVariables.BiotModulusInverse =
-            (rVariables.BiotCoefficient - rProp[POROSITY]) / rProp[BULK_MODULUS_SOLID] +
-            rProp[POROSITY] / rProp[BULK_MODULUS_FLUID];
+        // calculate Bulk modulus from stiffness matrix
+        return 1.0 - CalculateBulkModulus(rConstitutiveMatrix) / rProp[BULK_MODULUS_SOLID];
```

Comment:
Contents of CalculateBUlkModulus:
    return ConstitutiveMatrix(0, 0) - (4.0 / 3.0) * ConstitutiveMatrix(IndexG, IndexG);
The ConstitutiveMatrix is computed (relevant entries only)as:
    const Properties& r_material_properties = rValues.GetMaterialProperties();
    const auto        E                     = r_material_properties[YOUNG_MODULUS];
    const auto        NU                    = r_material_properties[POISSON_RATIO];
    const double c1 = (1.0 - NU) * c0;
    const double c3 = this->GetConsiderDiagonalEntriesOnlyAndNoShear() ? 0.0 : (0.5 - NU) * c0;
    C(INDEX_2D_PLANE_STRAIN_XX, INDEX_2D_PLANE_STRAIN_XX) = c1;
    C(INDEX_2D_PLANE_STRAIN_XY, INDEX_2D_PLANE_STRAIN_XY) = c3;

I strongly think that the input of CalculateBulkModulus should be rProp and no a constitutive matrix. That changed the argument lists of all functions calling CalculateBulkModulus. Further CalculateBulkModulus should be in a utility thing and functionality like CalculateBiotCoefficient as well. E.g. both in the transport equation utilities.

You may argue that it is out of scope for this PR, but the above indicates where I would like to end.

Reply:
I took the elastic constitutive thing, because I think that it is always the elastic matrix and not some plastic or otherwise nonlinear thing. That is why I'm hammering on taking it from material parameters. There are no objections against doing that in a different PR.